### PR TITLE
#164798904 Implement signup with JWT response

### PR DESCRIPTION
**Pull request by chukwudidu**
#### Description
Implement signup feature to enable users to get registered on the application. Upon successful signup, a token is registered and sent back.
#### Type of change
- New feature
Screenshot 2019-03-30 at 9 14 14 PM
Screenshot 2019-04-02 at 10 11 16 AM

#### Checklist:
- [x]  Write a test for signup route.
- [x]  Generate Token.
- [x] Hash password before saving to the database.
- [x] Add API documentation using swagger
#### PIVOTAL BOARD
[#164798904](https://www.pivotaltracker.com/story/show/164798904)

### CODE REVIEW

**Comment by BukkyOmo** [comment]
Good job. Check why travis is failing test.

**Comment by Chiazokam** [comment]
Hi Chuks, seems like you're returning only the token in the response. The user data (except the password) would also need to be returned.

I would add that right away

**Comment by chukwudidu** [comment]
> Good job. Check why travis is failing test.

Will do

File: server/utils/trycatch.helper.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
```
Comment (waleCloud) [inline server/utils/trycatch.helper.js]: @cupidcoder I'd like to know why this helper function exists and what it solves for?

File: server/test/auth/signup.spec.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import chai, { expect } from 'chai';
```
Comment (waleCloud) [inline server/test/auth/signup.spec.js]: I'd like to see more edge cases tested;

- negative testing
- make more comprehensive checks, aside from status code and token.

File: server/controllers/user.controller.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} responseObject
+   */
+  static async create(req, res) {
+    req.body.password = await passwordHash(req.body.password);
+    const newUser = await User.create(req.body);
+    const token = await generateToken({ id: newUser.id });
+    const { firstname, lastname, email } = newUser;
+    const userInfo = {
```
Comment (waleCloud) [inline server/controllers/user.controller.js:33]: You could do well to also destructure line `17` to avoid repetition.
a `try and catch` would suffice here, I know you did a `tryCatch` helper for this which I will be expecting some clarification on.
Good job!

File: server/utils/trycatch.helper.js (hunk 1 above)
Comment (chukwudidu) [reply server/utils/trycatch.helper.js]: @waleCloud the team agreed to use a try-catch wrapper that wraps all controller methods. This way, there wouldn't be repeated blocks of try-catch code.

File: server/test/auth/signup.spec.js (hunk 2 above)
Comment (chukwudidu) [reply server/test/auth/signup.spec.js]: @waleCloud I am not supposed to work on validating the input. Could you please shed more light on *negative testing* in this regard?

I would add further checks on user details like first name, last name and email

File: server/controllers/user.controller.js (hunk 3 above)
Comment (chukwudidu) [reply server/controllers/user.controller.js:33]: Will work on it right away

**Review by TheDrizzyWay — COMMENTED** [review]
Thanks for removing the repeated 'urlencoded' line.

File: server/utils/trycatch.helper.js (hunk 1 above)
Comment (waleCloud) [reply server/utils/trycatch.helper.js]: Okay cool.
I realized there's a status code of `50X` associated with the try catch.
My question is throwing a `50X` all the time the application fails. will be descriptive enough?
how are you handling a `40X` or related errors?

File: server/test/auth/signup.spec.js (hunk 2 above)
Comment (waleCloud) [reply server/test/auth/signup.spec.js]: Negative testing is a scenario where you supply wrong values/data inputs to the application or service you've built to see how well it handles it.

A wrong/invalid email, for example, should throw some kind of error.

Comment (waleCloud) [reply server/test/auth/signup.spec.js]: Why are you not suppose to work on validating the input?

File: server/controllers/user.controller.js (hunk 3 above)
Comment (waleCloud) [reply server/controllers/user.controller.js:33]: Hey, you destructured and still spelled out things on line `18`. 🤷‍♂️

File: server/utils/trycatch.helper.js (hunk 1 above)
Comment (chukwudidu) [reply server/utils/trycatch.helper.js]: We assumed that `40X` errors would be related to input validation errors which would be caught in the` middleware` validation functions @davidokonji would be writing.

File: server/test/auth/signup.spec.js (hunk 2 above)
Comment (chukwudidu) [reply server/test/auth/signup.spec.js]: Thanks for the clarification. There's a separate PT story for validating signup input and it's assigned to @davidokonji. You could confirm this here [#164798906](https://www.pivotaltracker.com/story/show/164798906)

File: server/controllers/user.controller.js (hunk 3 above)
Comment (chukwudidu) [reply server/controllers/user.controller.js:33]: Okay, I think I don't follow. Could you please explain further? 😬

File: package.json [hunk 4]
```diff
@@ -4,6 +4,7 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
+    "pretest": "cross-env NODE_ENV=test && cross-env NODE_ENV=test npm run migrate:undo && cross-env NODE_ENV=test npm run migrate",
```
Comment (paradisekelechi) [inline package.json:7]: Impressive

File: server/controllers/index.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import UserController from './user.controller';
+
+export {
+  // eslint-disable-next-line import/prefer-default-export
```
Comment (paradisekelechi) [inline server/controllers/index.js]: Why are we disabling this eslint rule?

File: server/controllers/user.controller.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+// eslint-disable-next-line import/named
```
Comment (paradisekelechi) [inline server/controllers/user.controller.js]: Why disable this eslint rule @cupidcoder ?

File: server/controllers/user.controller.js (hunk 3 above)
Comment (paradisekelechi) [reply server/controllers/user.controller.js:33]: After destructuring on line 17, you have a `password` variable. Is there any need to use `req.body.password`?
More so, since you plan to reassign `password`, don't you think you would have issues with `const`?

File: server/controllers/index.js (hunk 5 above)
Comment (chukwudidu) [reply server/controllers/index.js]: While following the agreed folder structure, we experienced issues using `export default` on that line and ESLint didn't allow only `export {..}`. That's why we resolved to disabling it

File: server/controllers/user.controller.js (hunk 6 above)
Comment (chukwudidu) [reply server/controllers/user.controller.js]: I will find a workaround this.

File: server/controllers/user.controller.js (hunk 3 above)
Comment (chukwudidu) [reply server/controllers/user.controller.js:33]: I will reimplement this.

File: server/controllers/index.js (hunk 5 above)
Comment (chukwudidu) [reply server/controllers/index.js]: I was able to find a workaround this.

File: server/test/signup.spec.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    it('should return token if user was successfully created', async () => {
+      const res = await chai.request(app)
+        .post('/api/v1/auth/signup')
+        .send(userMockData.validUserObject);
+
+      expect(res).to.have.status(201);
+      expect(res.body.data[0]).to.have.property('token');
+      expect(res.body.data[0].user).to.have.property('firstname').eql(userMockData.validUserObject.firstname);
+      expect(res.body.data[0].user).to.have.property('lastname').eql(userMockData.validUserObject.lastname);
+      expect(res.body.data[0].user).to.have.property('email').eql(userMockData.validUserObject.email);
```
Comment (paradisekelechi) [inline server/test/signup.spec.js:18]: Looks good
